src/OpenTelemetry.Instrumentation.Wcf/TelemetryPropagationReader.cs

```diff
@@ -0,0 +1,109 @@
+// <copyright file="TelemetryPropagationReader.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.ServiceModel.Channels;
+using OpenTelemetry.Internal;
+
+namespace OpenTelemetry.Instrumentation.Wcf;
+
+/// <summary>
+/// Pre-defined PropagationReader callbacks.
+/// </summary>
+public static class TelemetryPropagationReader
+{
+    private static readonly Func<Message, string, IEnumerable<string>> DefaultReader = Compose(HttpRequestHeaders, SoapMessageHeaders);
+
+    /// <summary>
+    /// Reads the values from the SOAP message headers. If the message is not a SOAP message it returns null.
+    /// </summary>
+    /// <param name="request">The incoming <see cref="Message"/> to read trace context from.</param>
+    /// <param name="name">The header name being requested.</param>
+    /// <returns>An enumerable of all the values for the requested header, or null if the requested header was not present on the request.</returns>
+    public static IEnumerable<string> SoapMessageHeaders(Message request, string name)
+    {
+        Guard.ThrowIfNull(request);
+        Guard.ThrowIfNull(name);
+
+        if (request.Version == MessageVersion.None)
+        {
+            return null;
+        }
+
+        var headerIndex = request.Headers.FindHeader(name, "https://www.w3.org/TR/trace-context/");
+        return headerIndex < 0
+            ? null
+            : new[] { request.Headers.GetHeader<string>(headerIndex) };
```

Comment:
This also instantiates a DataContractSerializer every time it's used. Either cache the DCS and use a different overload of GetHeader where you can provide the serializer, or call Headers.GetReaderAtHeader to get the XmlDictionaryReader for the header, then call ReadContentAsString to read the value. You might have to call `Read` to move forward in the state machine before you can call ReadContentAsString, but I'm not sure. You should wrap in a try/catch(XmlException) to be safe in case of a malformed header.